tools/extensions/validate_extensions.py

```diff
@@ -0,0 +1,238 @@
+    # Include all extensions from source/extensions/extensions_build_config.bzl
+    all_extensions = {}
+    all_extensions.update(extensions_build_config.EXTENSIONS)
+    for extension, target in all_extensions.items():
+        extension_db[extension] = get_extension_metadata(target)
+    # The TLS and generic upstream extensions are hard-coded into the build, so
+    # not in source/extensions/extensions_build_config.bzl
+    # TODO(mattklein123): Read these special keys from all_extensions.bzl or a shared location to
+    # avoid duplicate logic.
+    extension_db['envoy.transport_sockets.tls'] = get_extension_metadata(
+        '//source/extensions/transport_sockets/tls:config')
+    extension_db['envoy.upstreams.http.generic'] = get_extension_metadata(
+        '//source/extensions/upstreams/http/generic:config')
+    extension_db['envoy.upstreams.tcp.generic'] = get_extension_metadata(
+        '//source/extensions/upstreams/tcp/generic:config')
+    extension_db['envoy.upstreams.http.http_protocol_options'] = get_extension_metadata(
+        '//source/extensions/upstreams/http:config')
+    extension_db['envoy.request_id.uuid'] = get_extension_metadata(
+        '//source/extensions/request_id/uuid:config')
+    return extension_db
+
+
+def validate_extensions():
+    returns = 0
+    with open("source/extensions/extensions_metadata.yaml") as f:
+        metadata = yaml.safe_load(f.read())
+    returns += compare_old_and_new(generate_old_extension_db(), metadata)
+
+    if num_robust_to_downstream_network_filters(metadata) != num_read_filters_fuzzed():
+        returns += 1
+        print(
+            'Check that all network filters robust against untrusted'
+            'downstreams are fuzzed by adding them to filterNames() in'
+            'test/extensions/filters/network/common/uber_per_readfilter.cc')
+
+    for k, v in metadata.items():
+        if not v["security_posture"]:
+            returns += 1
+            print(
+                f"Missing security posture for {k}. "
+                "Please make sure the target is an envoy_cc_extension and security_posture is set")
+        elif v["security_posture"] not in EXTENSION_SECURITY_POSTURES:
+            print("Unknown extension security posture: {v['security_posture']}")
+            returns += 1
+        if not v["categories"]:
+            returns += 1
+            print(
+                f"Missing extension category for {k}. "
+                "Please make sure the target is an envoy_cc_extension and category is set")
+        else:
+            for cat in v["categories"]:
+                if cat not in EXTENSION_CATEGORIES:
+                    returns += 1
+                    print(f"Unknown extension category for {k}: {cat}")
+        if v["status"] not in EXTENSION_STATUS_VALUES:
+            returns += 1
+            print(f"Unknown extension status: {v['status']}")
+
+    return returns
+
+
+if __name__ == '__main__':
+    sys.exit(validate_extensions())
```

Comment:
Does this file need unit tests?

Reply:
yes - can i do it in the next PR when much of this code will be removed - im more than happy to shift this to a `Checker` class and add the appropriate tests